File: android/src/umich/framjack/FramingEngine.java
/*
 *  This file is part of hijack-infinity.
 *
 *  hijack-infinity is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  hijack-infinity is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with hijack-infinity.  If not, see <http://www.gnu.org/licenses/>.
 */

package umich.framjack;

import java.util.ArrayList;

public class FramingEngine {
	private IncomingPacketListener _incomingListener;
	private OutgoingByteListener _outgoingListener;
	
	private final int _receiveMax = 18;
	
	private enum ReceiveState { START, SIZE, DATA, DATA_ESCAPE};
	private ReceiveState _receiveState = ReceiveState.START;
	private ArrayList<Integer> _receiveBuffer;
	private int _receiveSize;
	
	private ArrayList<Integer> _transmitBuffer;
	
	public FramingEngine() {
		_receiveBuffer = new ArrayList<Integer>();
		_transmitBuffer = new ArrayList<Integer>();
	}
	
	public void receiveByte(int val) {
		//System.out.print(val + " ");
		if (val == 0xDD && _receiveState != ReceiveState.DATA_ESCAPE) {
			_receiveState = ReceiveState.SIZE;
			_receiveBuffer.clear();
			return;
		}
		
		switch (_receiveState) {
			case DATA:
				if (val == 0xCC) {
					_receiveState = ReceiveState.DATA_ESCAPE;
					_receiveSize--;
					break;
				}
				_receiveBuffer.add(val);
				if (_receiveBuffer.size() == _receiveSize) {
					checkPacket();
					_receiveState = ReceiveState.START;
				}
				break;
			case DATA_ESCAPE:
				_receiveState = ReceiveState.DATA;
				_receiveBuffer.add(val);
				if (_receiveBuffer.size() == _receiveSize) {
					checkPacket();
					_receiveState = ReceiveState.START;
				}
				break;
			case SIZE:
				if (val > _receiveMax) {
					_receiveState = ReceiveState.START;
					break;
				}
				_receiveSize = val;
				_receiveState = ReceiveState.DATA;
				break;
			default:
				break;
		}
	}
	
	private void checkPacket() {
		int sum = 0;
		for (int i = 0; i < _receiveBuffer.size() - 1; i++) {
			//System.out.print(Integer.toHexString(_receiveBuffer.get(i)) + " ");
			sum += _receiveBuffer.get(i);
		}
		//System.out.println(Integer.toHexString(_receiveBuffer.get(_receiveBuffer.size() - 1)));
		
		if ((sum & 0xFF) == _receiveBuffer.get(_receiveBuffer.size() - 1)) {
			int[] retArray = new int[_receiveBuffer.size() - 1];
			for (int i = 0; i < _receiveBuffer.size() - 1; i++) {
				retArray[i] = _receiveBuffer.get(i);
			}
			_incomingListener.IncomingPacketReceive(retArray);
		}		
	}
	
	public void transmitByte(int val) {
		if (val == 0xCC) {
			_transmitBuffer.add(0xDD);
		}
		_transmitBuffer.add(val);
	}
	
	public void transmitEnd() {
		int[] toSend = new int[_transmitBuffer.size() + 3];
		toSend[0] = 0xDD;
		toSend[1] = _transmitBuffer.size() + 1;
		for (int i = 0; i < _transmitBuffer.size(); i++) {
			toSend[i + 2] = _transmitBuffer.get(i);
			toSend[_transmitBuffer.size() + 2] += toSend[i + 2];
		}
		
		toSend[_transmitBuffer.size() + 2] &= 0xFF;
		_transmitBuffer.clear();
		_outgoingListener.OutgoingByteTransmit(toSend);
	}
	
	public void registerIncomingPacketListener(IncomingPacketListener listener) {
		_incomingListener = listener;
	}
	
	public void registerOutgoingByteListener(OutgoingByteListener listener) {
		_outgoingListener = listener;
	}
	
	public interface IncomingPacketListener {
		public abstract void IncomingPacketReceive(int[] packet);
	}
	
	public interface OutgoingByteListener {
		public abstract void OutgoingByteTransmit(int[] outgoingRaw); 
	}
}

File: hijack/src/framingEngine.c
/*
 *  This file is part of hijack-infinity.
 *
 *  hijack-infinity is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  hijack-infinity is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with hijack-infinity.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "framingEngine.h"

/////////////////////////////
// Region: Public functions
/////////////////////////////

void fe_init(void) {
	fe_state.rxState = fe_receiveState_start;
	fe_state.incomingBufferPos = 0;
	fe_state.outgoingBufferPos = 0;
	fe_state.outgoingBufferSize = 0;

	fe_state.incomingPktSize = 0;
	fe_state.outgoingPktSize = 0;

	fe_state.receiveSize = 0;
}

void fe_handleByteReceived(uint8_t val) {

	// If it's a start-byte abandon what we were
	// doing previously and reset state back to start.
	if (val == 0xDD &&
		fe_state.rxState != fe_receiveState_dataEscape) {
		fe_state.rxState = fe_receiveState_size;
		fe_state.incomingBufferPos = 0;
		return;
	}

	switch (fe_state.rxState) {
		// More than most likely it's data
		// or an escape character:
		case fe_receiveState_data:
			if (val == 0xCC) {
				fe_state.rxState = fe_receiveState_dataEscape;
				fe_state.receiveSize--;
				break;
			}
			// INTENTIONAL FALL THROUGH
		case fe_receiveState_dataEscape:
			fe_state.incomingBuffer[fe_state.incomingBufferPos++] = val;
			if (fe_state.incomingBufferPos == fe_state.receiveSize) {
				fe_checkPacket();
				fe_state.rxState = fe_receiveState_start;
			}
			break;

		// The byte after the start byte is the size:
		case fe_receiveState_size:
			if (val > FE_RECEIVEMAX) {
				fe_state.rxState = fe_receiveState_start;
				break;
			}
			fe_state.receiveSize = val;
			fe_state.rxState = fe_receiveState_data;
			break;
		default:
			break;
	}
}

void fe_handleByteSent(void) {
	// If we've sent the entire packet then lets call
	// the application layer and have it generate another
	// packet.
	if (fe_state.outgoingBufferPos == fe_state.outgoingBufferSize) {
		fe_state.packetSentCb();
		fe_buildTransmitBuffer();
		fe_state.outgoingBufferPos = 0;
	}

	fe_state.byteSender(fe_state.outgoingBuffer[fe_state.outgoingBufferPos++]);
}

void fe_writeTxBuffer(uint8_t * buf, uint8_t len) {
	uint8_t i;
	for (i = 0; i < len; i++) {
		fe_state.outgoingPktBuffer[i] = buf[i];
	}
	fe_state.outgoingPktSize = len;
}

void fe_startSending(void) {
	fe_buildTransmitBuffer();
	fe_handleByteSent();
}

//////////////////////////////////
// Region: Callback Subscriptions
//////////////////////////////////

void fe_registerPacketReceivedCb(fe_packetReceived * cb) {
	fe_state.packetReceivedCb = cb;
}

void fe_registerPacketSentCb(fe_callback * cb) {
	fe_state.packetSentCb = cb;
}

void fe_registerByteSender(fe_byteSender * sender) {
	fe_state.byteSender = sender;
}

////////////////////////////
// Region: Helper functions
////////////////////////////

void fe_checkPacket() {
	uint8_t sum = 0;
	uint8_t i = 0;
	uint8_t incomingPktPos = 0;

	// Compute the simple byte-addition checksum.
	for (i = 0; i < fe_state.incomingBufferPos - 1; i++) {
		sum += fe_state.incomingBuffer[i];
	}

	// Fill incomingPktBuffer with the data.
	if (sum == fe_state.incomingBuffer[fe_state.incomingBufferPos - 1]) {

	//	if ((LED_OUT & LED_0) == 0){
	//		LED_OUT ^= LED_0;
	//	} else {
	//		LED_OUT &= ~LED_0;
	//	}

		// Ignore the checksum on the end.
		for (i = 0; i < fe_state.incomingBufferPos - 1; i++) {
			fe_state.incomingPktBuffer[incomingPktPos++] =
				fe_state.incomingBuffer[i];
		}

		fe_state.incomingPktSize = incomingPktPos;

		fe_state.packetReceivedCb(fe_state.incomingPktBuffer,
			fe_state.incomingPktSize);
	}
}

void fe_buildTransmitBuffer(void) {
	// DEPENDS ON:
	// - outgoingPktBuffer being full
	// - outgoingPktSize being accurate

	// MODIFIES:
	// - outgoingBuffer
	// - outgoingBufferSize
	// - outgoingBufferPos

	uint8_t i;
	uint8_t sum = 0;

	fe_state.outgoingBufferPos = 0;
	fe_state.outgoingBuffer[fe_state.outgoingBufferPos++] = 0xDD;
	fe_state.outgoingBuffer[fe_state.outgoingBufferPos++] =
		fe_state.outgoingPktSize + 1;

	for (i = 0; i < fe_state.outgoingPktSize; i++) {

		// Properly escape invalid characters.
		if (fe_state.outgoingPktBuffer[i] == 0xDD ||
			fe_state.outgoingPktBuffer[i] == 0xCC) {
			fe_state.outgoingBuffer[fe_state.outgoingBufferPos++] = 0xCC;
			sum += 0xCC;
			fe_state.outgoingBuffer[1]++;
		}

		fe_state.outgoingBuffer[fe_state.outgoingBufferPos++] =
			fe_state.outgoingPktBuffer[i];

		sum += fe_state.outgoingPktBuffer[i];
	}

	fe_state.outgoingBuffer[fe_state.outgoingBufferPos++] = sum;
	fe_state.outgoingBufferSize = fe_state.outgoingBufferPos;
	fe_state.outgoingBufferPos = 0;
}
